modules/dreamview/backend/hmi/hmi.cc

```diff
@@ -0,0 +1,89 @@
+/******************************************************************************
+ * Copyright 2017 The Apollo Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *****************************************************************************/
+
+#include "modules/dreamview/backend/hmi/hmi.h"
+
+#include "modules/common/adapters/adapter_manager.h"
+#include "modules/common/util/util.h"
+#include "modules/dreamview/backend/common/dreamview_gflags.h"
+
+namespace apollo {
+namespace dreamview {
+
+using apollo::common::adapter::AdapterManager;
+
+HMI::HMI(WebSocketHandler *websocket) : websocket_(websocket) {
+  CHECK(common::util::GetProtoFromFile(FLAGS_hmi_config_filename, &config_))
+      << "Unable to parse HMI config file " << FLAGS_hmi_config_filename;
+
+  // Init status of modules and hardware.
+  for (const auto& module : config_.modules()) {
+    status_.add_modules()->set_name(module.name());
+  }
+  for (const auto& hardware : config_.hardware()) {
+    status_.add_hardware()->set_name(hardware.name());
+  }
+}
+
+void HMI::Start() {
+  AdapterManager::AddHMIStatusCallback(&HMI::OnHMIStatus, this);
+}
+
+void HMI::OnHMIStatus(const HMIStatus &hmi_status) {
+  // Note that we only merged status of modules and hardware, which matches the
+  // hmi_status_helper.
+  for (const auto& new_status : hmi_status.modules()) {
+    auto* old_status = GetModuleStatus(new_status.name());
+    if (old_status) {
+      old_status->MergeFrom(new_status);
```

Comment:
Can you just set the new status in the function instead of returning a pointer and call MergeFrom? not sure why there needs to be a MergeFrom rather than just setting the new status.

Reply:
MergeFrom is to support partial update.
Per offline discussion, we'll keep it for now and refactor later if necessary.